chancacher/chancacher.go

```diff
@@ -0,0 +1,365 @@
+// data to disk.
+package chancacher
+
+import (
+	"encoding/gob"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+// The maximum channel depth, which is also used when the channel depth is set
+// to 0. We could set this to MaxInt but we'd likely just run out of memory
+// without a clean way to triage. It's best to just enforce a sensible maximum.
+const MaxDepth = 1000000
+
+// A ChanCacher is a pipeline of channels with a variable-sized internal
+// buffer. The buffer can also cache to disk. The user is expected to connect
+// ChanCacher.In and ChanCacher.Out.
+type ChanCacher struct {
+	In      chan interface{}
+	Out     chan interface{}
+	runDone bool
+	maxSize int
+
+	cachePath      string
+	cache          bool
+	cacheR         *fileCounter
+	cacheW         *fileCounter
+	cacheEnc       *gob.Encoder
+	cacheModified  bool
+	cacheLock      sync.Mutex
+	cacheReading   bool
+	cachePaused    chan bool
+	cacheDone      chan bool
+	cacheAck       chan bool
+	cacheIsDone    bool
+	cacheCommitted bool
+}
+
+// Create a new ChanCacher with maximum depth, and optional backing file.  If
+// maxDepth == 0, the ChanCacher will be unbuffered. If maxDepth == -1, the
+// ChanCacher depth will be set to MaxDepth. To enable a backing store,
+// provide a path to backingPath. chancachers create two files using this
+// prefix named cache_a and cache_b.
+//
+// The maxSize argument sets the maximum amount of disk commit, in bytes.
+//
+// When a new ChanCacher is made, if cachePath points to existing cache files,
+// the ChanCacher will immediately attempt to drain them from disk. In this
+// way, you can recover data sent to disk on a crash or previous use of
+// Commit().
+func NewChanCacher(maxDepth int, cachePath string, maxSize int) (*ChanCacher, error) {
+	if cachePath != "" {
+		fi, err := os.Stat(cachePath)
+		if err != nil && !strings.Contains(err.Error(), "no such file or directory") {
+			return nil, err
+		}
+		if fi != nil && !fi.IsDir() {
+			return nil, fmt.Errorf("cache path not a directory")
+		}
+	}
+
+	// as close to infinite as possible...
+	if maxDepth == -1 || maxDepth > MaxDepth {
+		maxDepth = MaxDepth
+	}
+	c := &ChanCacher{
+		In:          make(chan interface{}),
+		Out:         make(chan interface{}, maxDepth),
+		cachePath:   cachePath,
+		cache:       cachePath != "",
+		cachePaused: make(chan bool),
+		cacheDone:   make(chan bool),
+		cacheAck:    make(chan bool),
+		maxSize:     maxSize,
+	}
+
+	// we start the cache unpaused, and because of go idioms, we have to
+	// make the channel in order for "closed" states to work - we can't
+	// just leave it initiated...
+	close(c.cachePaused)
+
+	go c.run()
+	if c.cache {
+		var err error
+
+		err = os.MkdirAll(c.cachePath, 0755)
+		if err != nil {
+			return nil, err
+		}
+
+		// create r and w files
+		r, err := os.OpenFile(filepath.Join(c.cachePath, "cache_a"), os.O_APPEND|os.O_CREATE|os.O_RDWR, 0644)
```

Comment:
This should merge the R and W files and create a single R file if W has data in it so we avoid new encoders against a single file.

Reply:
I think this is resolved by https://github.com/gravwell/gravwell/pull/1/commits/c4ec6e6e5fca1b669470beb6122421a061dc1cd9